### PR TITLE
Adds indexing operation to Metric class

### LINKED ISSUE #132: Implement __getitem__ as "metric arithmetic"
## 🚀 Feature

Allow a user to define a new metric that takes an item out of an other metric.

Basically :

```python
iou = IoU(num_classes=2, reduction="none")
fg_iou = iou[0]
bg_iou = iou[1]
```

### Motivation

There are multiple metrics (like IoU and confusion matrix) that would benefit from the use of such a feature, and it is close to the mechanism of metric arithmetic.

### Pitch

This would only need to define 
```python
class Metric:
    ...
    def __getitem__(self, idx):
        return CompositionalMetric(lambda x: x[idx], self, None)
```

### Alternatives

The straightforward alternative is to use CompositionalMetric directly.

### PR DESCRIPTION
**Pull request by victorjoos**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #132  (issue).

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/bases/test_composition.py [hunk 1]
```diff
@@ -304,7 +303,7 @@ def test_metrics_or(second_operand, expected_result):
 def test_metrics_pow(second_operand, expected_result):
     first_metric = DummyMetric(2)
 
-    final_pow = first_metric**second_operand
+    final_pow = first_metric ** second_operand
```
Comment (Borda) [inline tests/bases/test_composition.py:306]: this is yapf...

File: torchmetrics/metric.py [hunk 2]
```diff
@@ -127,7 +127,8 @@ def add_state(
                 If ``dist_reduce_fx`` is not callable or one of ``"mean"``, ``"sum"``, ``"cat"``, ``None``.
         """
         if (
-            not isinstance(default, Tensor) and not isinstance(default, list)  # noqa: W503
+            not isinstance(default, Tensor)
+            and not isinstance(default, list)  # noqa: W503
```
Comment (Borda) [inline torchmetrics/metric.py:131]:
```suggestion
            not isinstance(default, (Tensor, list))
```

File: torchmetrics/metric.py [hunk 3]
```diff
@@ -127,7 +127,8 @@ def add_state(
...
             or (isinstance(default, list) and len(default) != 0)  # noqa: W503
```
Comment (Borda) [inline torchmetrics/metric.py:132]:
```suggestion
            or (isinstance(default, list) and default)
```

File: torchmetrics/metric.py [hunk 4]
```diff
@@ -127,8 +127,8 @@ def add_state(
                 If ``dist_reduce_fx`` is not callable or one of ``"mean"``, ``"sum"``, ``"cat"``, ``None``.
         """
         if (
-            not isinstance(default, Tensor) and not isinstance(default, list)  # noqa: W503
-            or (isinstance(default, list) and len(default) != 0)  # noqa: W503
+            not isinstance(default, (Tensor, list))
+            or (isinstance(default, list) and default)
```
Comment (Borda) [inline torchmetrics/metric.py:131]:
```suggestion
            not isinstance(default, (Tensor, list)) or (isinstance(default, list) and default)
```

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Pull request merged**